src/flb_fstore.c

```diff
@@ -43,6 +43,108 @@ static int log_cb(struct cio_ctx *ctx, int level, const char *file, int line,
     return 0;
 }
 
+/* remove file from files_up list */
+static inline void files_up_remove(struct flb_fstore *fs,
+                            struct flb_fstore_file *fsf)
+{
+    if (fsf->_cache_head.next == NULL) {
+        return;
+    }
+
+    mk_list_del(&fsf->_cache_head);
+    --fs->files_up_counter;
+    if (fs->files_up_counter < 0) { /* remove when stable? */
+        flb_error("[fstore] fs->files_up_counter is negative which should never happen");
+    }
+}
+
+/*
+ * this function adds files to an fstore files_up list to track files that are
+ * mapped. the files_up list's order corresponds to how recent add is called on
+ * the files in the files_up list. recent files first, least recent files are last
+ * 
+ * max of max_chunks_up files allowed before eviction of least recently added file
+ */
+static struct flb_fstore_file *files_up_add(struct flb_fstore *fs,
+                         struct flb_fstore_file *fsf)
+{
+    struct flb_fstore_file *evict = NULL;
+
+    /* check if already in files_up */
+    if (fsf->_cache_head.next != NULL) {
+        /* send to front of list; renew */
+        files_up_remove(fs, fsf);
+        files_up_add(fs, fsf);
+        return NULL;
+    }
+ 
+    /* files_up cache full */
+    if (fs->files_up_counter >= fs->cio->max_chunks_up) {
+        /* evict oldest entry */
+        evict = mk_list_entry(fs->files_up.prev, struct flb_fstore_file, _cache_head);
+        files_up_remove(fs, evict);
+    }
+
+    /* add to files_up list, front */
+    mk_list_add_after(&fsf->_cache_head, &fs->files_up, &fs->files_up);
+    ++fs->files_up_counter;
+    return evict;
+}
+
+/* lets down file if needed, fsf must be brought up soon after by caller */
+static int file_up_prep(struct flb_fstore *fs, /* previously files_up_make_room */
+                         struct flb_fstore_file *fsf)
+{
+    struct flb_fstore_file *evict;
+    int ret;
+
+    evict = files_up_add(fs, fsf);
+    if (evict != NULL) {
+        ret = cio_chunk_down(evict->chunk);
+        if (ret != CIO_OK) {
+            flb_error("[fstore] error unmapping file chunk: %s:%s",
+                      fsf->stream->name, fsf->chunk->name);
+            return -1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * this function memory maps a file if it is unmapped, changing the file's state to up
+ * if it is down. if fsf->max_chunks_up files are up, a the least recently checked file
+ * will be evicted
+ * 
+ * files_up record is reordered to put recently checked files first.
+ */
+static int file_up_if_down(struct flb_fstore *fs,
```

Comment:
should we call file_mapped_if_down?

Reply:
IMO `_up_` should be used preferably to `_mapped_`, because we are interfacing with the `mmap` syscall through chunkio api. The chunkio concept of up for files is mmap, but it is probably a good idea for the code to follow chunkio's concept abstraction